### ISSUE TITLE
With nullable reference types enabled, the type of a var should be deduced as a non-nullable type if possible

### ISSUE THREAD

**Issue by firedraco** [dotnet/csharplang#3662]
Currently with nullable reference types, if I write:
```
var x = 2; //x is of type 'int'
```
but if I write:
```
var str = "test"; //str is of type 'string?'
```

This seems a little confusing as the type of var is not necessarily the same as the type of the expression on the right hand side. Someone could come in later and see that 'str' is a 'string?' and mistakenly think it was intentional.

I'd propose that if the right hand side of a var is a constant, a non-nullable reference, or value-type, then the deduced type of var should also be a non-nullable type (e.g. 'str' in the example above should be of type 'string').

This does require the programmer to type out the entire type should they desire a nullable version of a type, but this already occurs if you want to store a class-typed result into an interface, for example, so I do not believe that is a major issue. Perhaps we could introduce a new syntax of `var?` to mean a deduced type but nullable, but I am unsure of the impact on the compiler etc. that that suggestion would introduce.

**Comment by HaloFour** [comment]
The team already considered `var?` and decided that treating `var` as nullable but that it uses flow analysis to determine nullability at that point was the best compromise when it comes to null-safety vs. having to modify existing code.

**Comment by 333fred** [comment]
See the previous discussion here: https://github.com/dotnet/csharplang/issues/3591#issue-642469895

TLDR: This behavior is intentional. `var` means "Figure out the type for me", not "Figure out the type for me except for nullability." Our dogfooding experience and early feedback revealed that rather than give `var` more switches that need to be added after the fact, we should just allow `var` to be assigned any value. The state is still tracked, and you still get warnings, but you don't have to go back and change types of variables because you changed your code, which is `var`'s intended purpose.

**Comment by firedraco** [comment]
I had much the same opinion as LWChris in that thread; my understanding was that non-nullable references would be the default if the toggle was on, and not if the toggle was off, but it seems like nullable references are still the default in both cases so existing code doesn't have to change (which makes sense).

I'll go for the analyzer route then, thanks!

**Comment by HaloFour** [comment]
Note that even if the variable is considered nullable the current nullability is inferred from the assignment, so if you are assigning it a non-null reference the variable is non-null and will not warn on dereference.  The only practical difference is that you can now assign a maybe-null value to that variable without getting a warning, but any attempt to dereference that variable afterwards without first doing a null check will warn.

**Comment by jnm2** [comment]
The IDE currently makes this very confusing. It catches me out several times a week. When you hover over the `str` identifier in the `var str = ` line, it shows a nullable type and does not show "not null at this location" like it does if you hover over subsequent usages of the identifier.

**Comment by CyrusNajmabadi** [comment]
@jnm2 Yes, this is a known issue.  I believe @jasonmalinowski might be working on a fix here.

**Comment by jnm2** [comment]
I think even the normal "May be null/not null here" display isn't emphasized strongly enough relative to the nullable type displayed first. Could "not null" be bold when the type is nullable, or always?

![image](https://user-images.githubusercontent.com/8040367/87191610-c4d5a080-c2c2-11ea-96e8-2f1bf0c7cd7b.png)

Conversely, could "maybe null" be bold when the type is non-nullable, or always?

![image](https://user-images.githubusercontent.com/8040367/87191699-ee8ec780-c2c2-11ea-99ec-b33ce3efbaed.png)

**Comment by CyrusNajmabadi** [comment]
I'm open to that being experimented with.  @jasonmalinowski  ^

**Comment by jasonmalinowski** [comment]
@ryzngard is the one looking at these bugs: https://github.com/dotnet/roslyn/issues/41245

**Comment by theunrepentantgeek** [comment]
@jnm2 I'd suggest that bold (or other emphasis) should be added to only one of the messages, to make them more visually distinct - and since the **may be null** case is the one which might indicate there's a bug, that's the one I'd emphasize.

**Comment by ryzngard** [comment]
Keep in mind assignment always shows the previous state as well. Regardless of how we emphasize the wording, this is how the flow state works.

Compare these two images, note that hover is approximately above where dialog is shown. I was taking quick screenshots so my cursor isn't showing there.

## Assignment To Null
![image](https://user-images.githubusercontent.com/475144/87202909-60144900-c2b6-11ea-9e62-4c4ef546db31.png)

## After Assignment To Null
![image](https://user-images.githubusercontent.com/475144/87202934-6acede00-c2b6-11ea-979b-7dfb4ca79796.png)

**Comment by ryzngard** [comment]
For providing emphasis on wording, I think this is a separate issue.  I filed #45888 to track how we could better show emphasis. I like the idea of bold on "may be null" like @theunrepentantgeek  suggests

**Comment by jnm2** [comment]
It's vanishingly rare for "may be null" to conflict with the type because there would be a warning in code if so. The thing I keep tripping over is that I see the variable type with the `?` and the "not null" text doesn't catch my eye.

**Comment by HaloFour** [comment]
@ryzngard 

> Keep in mind assignment always shows the previous state as well.

hm, that kinda makes sense, but I don't think that's intuitive at all.  If the value is the target of an assignment and not a part of an expression IMO it would make more sense to reflect the nullability state as it will be after the assignment.  And that tells me nothing about the nullability state of the local at the initial assignment, which is where I'm most likely to hover to try to determine the type of a variable.
